### PR TITLE
fast/mediastream/getDisplayMedia-max-constraints5.html is flakily crashing

### PR DESCRIPTION
**Pull request by szewai**
#### e11ba9225da49ccf8e41a68b5f101c424a0aa22c
<pre>
fast/mediastream/getDisplayMedia-max-constraints5.html is flakily crashing
<a href="https://bugs.webkit.org/show_bug.cgi?id=285249">https://bugs.webkit.org/show_bug.cgi?id=285249</a>
<a href="https://rdar.apple.com/142189100">rdar://142189100</a>

Reviewed by Chris Dumez.

Crash trace indicates reference to a MockRealtimeVideoSource object is created on a background thread, while the object
is being destroyed on the main thread. The cause is MockRealtimeVideoSource is created and destroyed on main runloop,
while m_emitFrameTimer is created with raw pointer of MockRealtimeVideoSource and non-main runloop (m_emitFrameTimer
will create reference to the object when it is fired on the non-main runloop, see 286015@main). To fix this, make
m_emitFrameTimer capture ThreadSafeWeakPtr instead.

* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::m_emitFrameTimer):

Canonical link: <a href="https://commits.webkit.org/288344@main">https://commits.webkit.org/288344@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/3db3905311c1db8054a0e077f5732ce837e06e56

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82676 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2327 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36831 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87796 "Hash 3db39053 for PR 38432 does not build (failure)") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33734 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84770 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2399 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10238 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64419 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22186 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85733 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1744 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75205 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44694 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1639 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29391 "Passed tests") | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32768 "Hash 3db39053 for PR 38432 does not build (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72916 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30051 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89162 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9969 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7221 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72834 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10197 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71014 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72048 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16197 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15121 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1358 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12811 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9922 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15443 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9796 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13261 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11565 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp [hunk 1]
```diff
@@ -781,6 +781,18 @@ void MockRealtimeVideoSource::endApplyingConstraints()
     m_beingConfigured = false;
 }
 
+RunLoop::Timer& MockRealtimeVideoSource::ensureEmitFrameTimer()
+{
+    if (!m_emitFrameTimer) {
+        m_emitFrameTimer = makeUnique<RunLoop::Timer>(m_runLoop.get(), [weakThis = ThreadSafeWeakPtr { *this }]() {
```
Comment (cdumez) [inline Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp]: Why do we need to delay the construction of the Timer? Why couldn't we still construct the timer in the constructor?

Comment (szewai) [reply Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp]: I reverted this change.
tl;dr: I remembered we are not allowed to call `ref()` in constructor, and it seems we could create timer when it's needed, so I moved the creation out. It seems creating `ThreadSafeWeakPtr` only invokes `weakRef()`, so it should be fine.

Comment (cdumez) [reply Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp]: You're allowed to ref in the constructor, just call `relaxAdoptionRequirement()` if needed.

**Review by cdumez — APPROVED** [review]
LGTM
